rust/lib.rs

```diff
@@ -1,8 +1,92 @@
-#[derive(PartialEq, Debug)]
-pub struct Schema {}
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::fs::{read_to_string, File};
 
-pub fn get_schema(topic: &str, version: Option<u16>) -> Option<Schema> {
-    None
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[non_exhaustive]
+enum SchemaType {
+    #[serde(rename = "json")]
+    Json,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[non_exhaustive]
+enum CompatibilityMode {
+    #[serde(rename = "none")]
+    None,
+    #[serde(rename = "backward")]
+    Backward,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum SchemaError {
+    TopicNotFound,
+    InvalidVersion,
+    InvalidType,
+}
+
+impl fmt::Display for SchemaError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SchemaError::TopicNotFound => write!(f, "Topic not found"),
+            SchemaError::InvalidVersion => write!(f, "Invalid version"),
+            SchemaError::InvalidType => write!(f, "Invalid type"),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct TopicSchema {
+    version: u16,
+    #[serde(rename = "type")]
+    schema_type: SchemaType,
+    compatibility_mode: CompatibilityMode,
+    resource: String,
+    examples: Vec<String>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct TopicData {
+    topic: String,
+    schemas: Vec<TopicSchema>,
+}
+
+impl TopicData {
+    fn load(path: &str) -> Result<Self, SchemaError> {
+        let f = File::open(path).map_err(|_| SchemaError::TopicNotFound)?;
+        serde_yaml::from_reader(f).map_err(|_| SchemaError::TopicNotFound)
+    }
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct Schema {
+    version: u16,
+    schema_type: SchemaType,
+    compatibility_mode: CompatibilityMode,
+    // TODO: Actually parse the json schema
+    schema: String,
+}
+
+pub fn get_schema(topic: &str, _version: Option<u16>) -> Result<Schema, SchemaError> {
+    let topic_path = format!("./topics/{}.yaml", topic);
+    let topic_data = TopicData::load(&topic_path)?;
+    // TODO: Respect version
+    let latest = topic_data.schemas.last().unwrap();
+    if latest.schema_type != SchemaType::Json {
+        return Err(SchemaError::InvalidType);
+    }
+
+    let json_schema_path = format!("./schemas/{}", latest.resource);
+
+    Ok({
+        Schema {
+            version: latest.version,
+            schema_type: latest.schema_type.clone(),
+            compatibility_mode: latest.compatibility_mode.clone(),
```

Comment:
Not sure if there's an easy way to avoid these clones

Reply:
add `derive(Clone)` to the CompatibilityMode enum, but i think you should be able to remove the `.clone()` already anyway? It will mean you can't use `latest` afterwards but you're returning anyway

Reply:
ah, you need to replace `last()` with something like `pop()` to make `latest` of type `Schema` instead of `&Schema`